src/tree.py

```diff
@@ -0,0 +1,168 @@
+            index = w.indexlist
+            edge_l = index[1]-index[0]
+            lx = len(x)
+            substr_len = lx-i-1
+
+
+            # Match through edge
+            for match_l_edge in range(1,edge_l): # just edge or edge+1?
+                same = x[index[0]+match_l_edge] == x[i+match_l_edge]
+
+                # matched until end of string, not end of edge
+                if same and match_l_edge == substr_len: # and substr_len < edge_l: #do we need this?
+                    return w, match_l_edge, match_length_suf+match_l_edge
+                elif not same:
+                    return w, match_l_edge, match_length_suf+match_l_edge
+
+            # search the next node, update how far we have matched so far
+            return search_node(x, w, i+edge_l, match_length_suf+edge_l)
+
+        else:
+            return node, 0, match_length_suf #i-1
+    
+    root = T[0]
+    return search_node(x, root, i, 0)
+
+def extend_from_node(node, i, match_l_suf, x, T):
+    # parent is node
+    # create leaf node with string[i:]
+
+    lx = len(x)
+    index_list = [match_l_suf+i, lx]
+    new_node = Node(index_list, None, i)
+    T.append(new_node)
+
+    node.reproduction(new_node, x[index_list[0]]) # x[i]
+
+    return T
+
+def extend_from_edge(node, i, match_l_suf, x, T, match_l):
+
+    # create new node w = node[:match_l] with parent of 'node' as parent, string and node as children
+    # update parent child from node to w
+    # update node indexlist to [match_l:], change parent to w
+
+    previous_indexes = node.indexlist
+    x_mismatch_letter = x[previous_indexes[0]+match_l]
+
+    # create the intermediate node
+    w_index_list = [previous_indexes[0], previous_indexes[0]+match_l]       
+    w = Node(w_index_list, None, {}) #node.parent doesnt exist 
+
+    # make the parent of node the parent of w, and w its child
+    node.parent.reproduction(w, x[previous_indexes[0]])
+    T.append(w)
+
+    # create the new node
+
+    lx = len(x)
+    new_index_list = [i+match_l_suf, lx] #[match_l_suf+i, lx] # how long we matched the suffix w edge +1
+    new_node = Node(new_index_list, None, i) # index list, parent, label because leaf
+    T.append(new_node)
+
+    # change node index
+    node.update_index(0, previous_indexes[0]+match_l)
+
+    # add w children
+    w.reproduction(new_node, x[new_index_list[0]]) # string[match_l]
+    w.reproduction(node, x_mismatch_letter)
+
+    return T
+
+def subtree_labels(current_n):
+        if current_n.is_leaf():
+            yield current_n.children
+        else:
+            for v in current_n.children.values():
+                yield from subtree_labels(v)
+
+def Convert(lst):
+    res_dct = {lst[i][0]: lst[i][1] for i in range(0, len(lst))}
+    return res_dct
+
+def sorted_tree(T):
+    
+    for branch in T:
+        if type(branch.children) == dict:
+            
+            branch.children = Convert(sorted(branch.children.items(), key=lambda item: item[0]))
```

Comment:
```python
branch.children = {key: branch.children[key] for key in sorted(branch.children.keys())}
```

if you absolutely want to save children in dictionaries and want them sorted as well. It costs $O(\sigma\log\sigma)$ per node.

If you only want this dict sorted to output the leaves sorted, you could of course also make _one_ dictionary that maps letters to their rank:

```python
rank = {a: i for i, a in enumerate(sorted(set(x)))}
```

At any point where you need the children sorted, you can then do

```python
children = [None] * len(rank)
for a, child in branch.children.items():
    children[rank[a]] = child
```

and now the children are in a sorted list (with `None` when there is a letter missing). That only takes time $O(\sigma)$ per node.